crates/qpy/src/py_methods.rs

```diff
@@ -0,0 +1,666 @@
+            };
+            return Ok(Some(new_name));
+        }
+
+        if ["ControlledGate", "AnnotatedOperation"].contains(&name.as_str())
+            || is_python_gate(py, op, imports::MCMT_GATE.get_bound(py))?
+        {
+            return Ok(Some(format!("{}_{}", op.name(), Uuid::new_v4())));
+        }
+
+        if is_python_gate(py, op, imports::PAULI_EVOLUTION_GATE.get_bound(py))? {
+            return Ok(Some(format!("###PauliEvolutionGate_{}", Uuid::new_v4())));
+        }
+
+        Ok(None)
+    })
+}
+
+/// when trying to instantiate nonstandard gates, we turn to the relevant python clas
+/// this function obtains the class based on the gate class name
+pub fn get_python_gate_class<'a>(
+    py: Python<'a>,
+    gate_class_name: &String,
+) -> PyResult<Bound<'a, PyAny>> {
+    let library = py.import("qiskit.circuit.library")?;
+    let circuit_mod = py.import("qiskit.circuit")?;
+    let control_flow = py.import("qiskit.circuit.controlflow")?;
+    if library.hasattr(gate_class_name)? {
+        library.getattr(gate_class_name)
+    } else if circuit_mod.hasattr(gate_class_name)? {
+        circuit_mod.getattr(gate_class_name)
+    } else if control_flow.hasattr(gate_class_name)? {
+        control_flow.getattr(gate_class_name)
+    } else if gate_class_name == "Clifford" {
+        Ok(imports::CLIFFORD.get_bound(py).clone())
+    } else if gate_class_name == "pauli_product_measurement" {
+        Ok(imports::PAULI_PRODUCT_MEASUREMENT.get_bound(py).clone())
+    } else {
+        Err(PyIOError::new_err(format!(
+            "Gate class not found: {:?}",
+            gate_class_name
+        )))
+    }
+}
+
+// serializes python metadata to JSON using a python JSON serializer
+pub fn serialize_metadata(
+    metadata_opt: &Option<Bound<PyAny>>,
+    metadata_serializer: Option<&Bound<PyAny>>,
+) -> PyResult<Bytes> {
+    match metadata_opt {
+        None => Ok(Bytes::new()),
+        Some(metadata) => {
+            let py = metadata.py();
+            let none = py.None();
+            let py_serializer = metadata_serializer.unwrap_or(none.bind(py));
+            let json = py.import("json")?;
+            let kwargs = PyDict::new(py);
+            kwargs.set_item("separators", PyTuple::new(py, [",", ":"])?)?;
+            kwargs.set_item("cls", py_serializer)?;
+            Ok(json
+                .call_method("dumps", (metadata,), Some(&kwargs))?
+                .extract::<String>()?
+                .into())
+        }
+    }
+}
+
+// helper method to extract attribute from a py_object
+pub fn getattr_or_none<'py>(
+    py_object: &'py Bound<'py, PyAny>,
+    name: &str,
+) -> Option<Bound<'py, PyAny>> {
+    match py_object.getattr(name) {
+        Ok(val) => {
+            if val.is_none() {
+                None
+            } else {
+                Some(val)
+            }
+        }
+        Err(_) => None,
+    }
+}
+
+pub fn py_serialize_numpy_object(py_object: &Py<PyAny>) -> PyResult<Bytes> {
```

Comment:
We use the `py_` prefix for functions which are accessible through PyO3. I would recommend not adapting this convention to avoid a confusion (this is relevant for other functions in this file).

Reply:
As I wrote in the other comment, I would go with an explicit `Python` parameter.